code/go/0chain.net/smartcontract/dbs/event/user_aggregate.go

```diff
@@ -19,99 +19,244 @@ type UserAggregate struct {
+		for _, u := range *users {
+			updatedAggrs = append(updatedAggrs, UserAggregate{
+				Round:     event.BlockNumber,
+				UserID:    u.UserID,
+				PayedFees: u.PayedFees,
+			})
+		}
 		return
-	}
-	logging.Logger.Debug("getting user aggregate ids", zap.Int("num", len(ids)))
+	},
+	TagUpdateUserCollectedRewards: func(event Event) (updatedAggrs []UserAggregate) {
+		users, ok := fromEvent[[]UserAggregate](event.Data)
+		if !ok {
+			logging.Logger.Error("user aggregate",
+				zap.Any("event", event.Data), zap.Error(ErrInvalidEventData))
+			return
+		}
+		logging.Logger.Debug("user_aggregates TagUpdateUserCollectedRewards", zap.Int("events", len(*users)))
+		for _, u := range *users {
+			updatedAggrs = append(updatedAggrs, UserAggregate{
+				Round:           event.BlockNumber,
+				UserID:          u.UserID,
+				CollectedReward: u.CollectedReward,
+			})
+		}
+		return
+	},
+}
 
-	var currentUsers []User
-	result := edb.Store.Get().Model(&User{}).
-		Where("users.user_id in (select id from temp_user_ids ORDER BY id limit ? offset ?)", limit, offset).
-		Find(&currentUsers)
+func (edb *EventDb) GetLatestUserAggregates(ids map[string]interface{}) (map[string]*UserAggregate, error) {
+	var ua []UserAggregate
+	var mappedAggrs = make(map[string]*UserAggregate, len(ua))
+
+	var idlist []string
+	for id := range ids {
+		idlist = append(idlist, id)
+	}
 
+	if len(idlist) == 0 {
+		logging.Logger.Info("empty aggregates list")
+		return mappedAggrs, nil
+	}
+	result := edb.Store.Get().
+		Raw(`SELECT user_id, max(round), collected_reward, payed_fees, total_stake, read_pool_total, write_pool_total 
+	FROM user_aggregates 
+	WHERE user_id IN (SELECT unnest(?::text[]))
+	GROUP BY user_id, collected_reward, payed_fees, total_stake, read_pool_total, write_pool_total`, pq.Array(idlist)).
+		Scan(&ua)
 	if result.Error != nil {
-		logging.Logger.Error("getting current users", zap.Error(result.Error))
-		return
+		logging.Logger.Error("can't select aggregates", zap.Error(result.Error))
+		return nil, result.Error
 	}
-	logging.Logger.Debug("user_snapshot", zap.Int("total_current_users", len(currentUsers)))
 
-	if round <= edb.AggregatePeriod() && len(currentUsers) > 0 {
-		if err := edb.addUserSnapshot(currentUsers); err != nil {
-			logging.Logger.Error("saving users snapshots", zap.Error(err))
+	for _, aggr := range ua {
+		u := aggr
+		mappedAggrs[u.UserID] = &u
+	}
+
+	return mappedAggrs, nil
+}
+
+func (edb *EventDb) updateUserAggregates(e *blockEvents) error {
+	logging.Logger.Debug("calculating user aggregates", zap.Int64("round", e.round))
+	var updatedAggrs []UserAggregate
+	for _, ev := range e.events {
+		if h, ok := handlers[ev.Tag]; ok {
+			aggrs := h(ev)
+			updatedAggrs = append(updatedAggrs, aggrs...)
 		}
 	}
 
-	oldUsers, err := edb.getUserSnapshots(limit, offset)
+	ids := make(map[string]interface{})
+	for _, aggr := range updatedAggrs {
+		aggr.Round = e.round
```

Comment:
`aggr` is value, not pointer, `aggr.Round = e.round` would not change it actually.